CIScripts/Common/Invoke-NativeCommand.Tests.ps1

```diff
@@ -0,0 +1,121 @@
+            $Command = "whoami.exe"
+            (Invoke-NativeCommand -CaptureOutput -AllowNonZero { & $Command }).ExitCode | Should Be 0
+        }
+
+        It "preserves the ErrorActionPreference" {
+            $ErrorActionPreference = "Stop"
+            Invoke-NativeCommand -CaptureOutput { whoami.exe } | Out-Null
+            $ErrorActionPreference | Should Be "Stop"
+        }
+    }
+
+    Context "Remote machine" {
+        BeforeAll {
+            $Sessions = New-RemoteSessions -VMs (Read-TestbedsConfig -Path $TestenvConfFile)
+            [
+                Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSUseDeclaredVarsMoreThanAssignments",
+                "Session",
+                Justification="PSAnalyzer doesn't understand relations of Pester's blocks.")
+            ]
+            $Session = $Sessions[0]
+        }
+
+        It "does not throw on successful command" {
+            Invoke-NativeCommand -Session $Session { whoami.exe }
+        }
+
+        It "can accept the scriptblock also as an explicit parameter" {
+            Invoke-NativeCommand -Session $Session -CaptureOutput -ScriptBlock { whoami.exe }
+        }
+
+        It "throws on failed command" {
+            { Invoke-NativeCommand -Session $Session { whoami.exe /invalid_parameter } } | Should Throw
+        }
+
+        It "throws on nonexisting command" {
+            { Invoke-NativeCommand -Session $Session { asdfkljasdsdf.exe } } | Should Throw
+        }
+
+        It "captures the exitcode of successful command" {
+            $Result = Invoke-NativeCommand -Session $Session -AllowNonZero { whoami.exe }
+            $Result.ExitCode | Should Be 0
+        }
+
+        It "captures the exitcode a failed command" {
+            (Invoke-NativeCommand -Session $Session -AllowNonZero { whoami.exe /invalid }).ExitCode | Should Not Be 0
+        }
+
+        It "can capture the output of a command" {
+            (Invoke-NativeCommand -Session $Session -CaptureOutput { whoami.exe }).Output | Should BeLike '*\*'
+        }
+
+        It "can capture multiline output" {
+            (Invoke-NativeCommand -Session $Session -CaptureOutput { whoami.exe /? }).Output.Count | Should BeGreaterThan 1
+        }
+
+        It "does not capture the output by default" {
+            Invoke-NativeCommand -Session $Session { whoami.exe } | Should BeNullOrEmpty
+        }
+
+        It "can use variables it scriptblock" {
+            $Command = "whoami.exe"
+            (Invoke-NativeCommand -Session $Session -CaptureOutput -AllowNonZero { & $Using:Command }).ExitCode | Should Be 0
+        }
+
+        It "preserves the ErrorActionPreference" {
+            $OldEA = Invoke-Command -Session $Session { $ErrorActionPreference }
+            Invoke-NativeCommand -Session $Session -CaptureOutput { whoami.exe } | Out-Null
+            Invoke-Command -Session $Session { $ErrorActionPreference } | Should Be $OldEA
+        }
+    }
+}
```

Comment:
Huh, I'm really surprised, because I'm quite convinced I've tried that before and was pretty sure that variables from outside the `It` block are invisible inside it. (Anyway, the proper solution is to use `-TestCases` parameter)

Unfortunately, parametrizing the tests somehow with
```powershell
$TestCases = @(
    @{ MaybeSession = $null }
    @{ MaybeSession = $Session }
)
```
require `-Session $MaybeSession` parameter on `Invoke-NativeCommand` to silently accept a `$null` session. Do we wan't that?

> maybe let's pass a string and then build each test case like a string?

I was tempted to say *Not on* my *watch*, but I have quite a crazy idea – accept a scriptblock with a placeholder flag `-MaybeSession`, stringify, replace with `"-Session $Session"` or `""`, and then convert back to a scriptblock.

Anyway, I think that tests should be as simple as possible and there should be at least a few "regular" tests.

Reply:
I've just tried such a magic:

```powershell
    function ItLocalAndRemote {
        Param(
            [String] $Name,
            [ScriptBlock] $ScriptBlock
        )

        function ReplaceInScriptBlock {
            Param(
                [String] $Placeholder,
                [String] $With,
                [Parameter(ValueFromPipeline = $true)] [ScriptBlock] $ScriptBlock
            )
            [String] $ScriptString = $ScriptBlock
            [ScriptBlock]::Create(($ScriptString -Replace $Placeholder, $With))
        }

        $Placeholder = '-MaybeSession'
        It "local: $Name" ($ScriptBlock | ReplaceInScriptBlock $Placeholder -With '')
        It "remote: $Name"  ($ScriptBlock | ReplaceInScriptBlock $Placeholder -With '-Session $Session')
    }

...

        ItLocalAndRemote "works on a simple case" {
            Invoke-NativeCommand { whoami.exe }
            Get-WriteHostOutput | Should BeLike '*\*'
        }
```

And it almost worked...

```
        CommandNotFoundException: The term 'Invoke-NativeCommand' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
        at <ScriptBlock>, <No file>: line 2
```

(It seems that imported functions are not visible in such a scriptblock (eg. regular `Invoke-Command` worked))

Also, the error reporting is worse in this case, as you can see above.

Reply:
I commented about this approach in a comment above:

> require -Session $MaybeSession parameter on Invoke-NativeCommand to silently accept a $null session. Do we wan't that?

Do think we should enable passing `$null` as session to `Invoke-NativeCommand`? (Parameters by default don't accept null)

Reply:
It would simplify tests a lot and could make potential client code more "generic". I'd say yes

Reply:
@krdln actually, maybe we could use closure mechanism here as well, instead? (as discussed in another comment chain)

Reply:
A `Make-InvokeNativeCommand` function that would return a closure with or without using `-Session`